### PR TITLE
Refactor parfor: split up ParforPass

### PR DESCRIPTION
**Pull request by sklam**
* Split up the `ParforPass` into sub-passes that does specific transformations.
* Add tests for each of the sub-passes. These tests are not exhaustive for each transformations yet, but they produces a good code-coverage to ease further refactoring.

### CODE REVIEW

File: numba/parfors/parfor.py [hunk 1]
```diff
@@ -1403,6 +1409,7 @@ def replace_func():
                                         break
                             return True
                         if guard(replace_func):
+                            self.stats['replaced_func'] += 1
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/parfors/parfor.py]: That's already done by: `swapped` in https://github.com/numba/numba/blob/f9c27dd8357da8a3674bcaa9c2f6dde9c257b02d/numba/parfors/parfor.py#L1408

[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/parfors/parfor.py]: and... I actually wrote that, I remember now.

File: numba/parfors/parfor.py [hunk 2]
```diff
@@ -1684,107 +1517,602 @@ def _convert_setitem(self, blocks):
...
+        typemap[tuple_var.name] = types.containers.UniTuple(
+            types.uintp, ndims)
+        tuple_call = ir.Expr.build_tuple(list(index_vars), loc)
+        tuple_assign = ir.Assign(tuple_call, tuple_var, loc)
+        body_block.body.append(tuple_assign)
+        return tuple_var, types.containers.UniTuple(types.uintp, ndims)
+    elif ndims == 1:
+        return index_vars[0], types.uintp
+    else:
+        raise NotImplementedError(
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/parfors/parfor.py]: I didn't change this. I don't want to break tests yet. But, in general, yes, we should change this sort of things everywhere in this file.

File: numba/parfors/parfor.py [hunk 3]
```diff
@@ -1846,8 +2174,11 @@ def find_mask_from_size(size_var):
                         # input array has to be dead after loop
                         require(arr_var.name not in live_vars)
                         # loop for arr's definition, where size = arr.shape
-                        arr_def = get_definition(self.func_ir, size_def.value)
-                        result = self._find_mask(arr_def)
+                        arr_def = get_definition(pass_states.func_ir, size_def.value)
+                        result = _find_mask(pass_states.typemap, pass_states.func_ir, arr_def)
+
+                        # The following is never tested.
+                        raise AssertionError("unreachable")
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/parfors/parfor.py]: oh the `guard`, i have to check that.

File: numba/tests/test_parfors_passes.py [hunk 4]
```diff
@@ -0,0 +1,647 @@
...
+        self.state.typemap = None
+        self.state.return_type = None
+        self.state.calltypes = None
+
+
+class BaseTest(TestCase):
+    @classmethod
+    def _run_parfor(cls, test_func, args):
+        # TODO: refactor this with get_optimized_numba_ir() where this is
+        #       copied from
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/tests/test_parfors_passes.py]: Needs to do this when Parfors become proper `FunctionPass`

File: numba/parfors/parfor.py [hunk 5]
```diff
@@ -1462,20 +1469,16 @@ def replace_func():
...
             return ft
 
-class ParforPass(object):
 
-    """ParforPass class is responsible for converting Numpy
-    calls in Numba intermediate representation to Parfors, which
-    will lower into either sequential or parallel loops during lowering
-    stage.
-    """
+class ParforPassStates:
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/parfors/parfor.py]: added in https://github.com/numba/numba/pull/5272/commits/ee1385ecf9efd6fea78ff3a2bf21731203e99491

File: numba/parfors/parfor.py [hunk 6]
```diff
@@ -1490,191 +1493,21 @@ def __init__(self, func_ir, typemap, calltypes, return_type, typingctx,
...
-                        or callname == ('reduce', '_functools')):
-                        # reduce function with generic function
-                        parfor = guard(self._reduce_to_parfor, equiv_set, lhs,
-                                       expr.args, loc)
-                    if parfor:
-                        instr = parfor
-                new_body.append(instr)
-            block.body = new_body
-        return
+class ConvertSetItemPass:
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/parfors/parfor.py]: added in https://github.com/numba/numba/pull/5272/commits/ee1385ecf9efd6fea78ff3a2bf21731203e99491

File: numba/parfors/parfor.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sklam) [reply numba/parfors/parfor.py]: guard only capture `GuardException`

File: numba/tests/test_parfors.py [hunk 7]
```diff
@@ -43,7 +43,7 @@
 import unittest
 
 
-test_disabled = unittest.skipIf(True, 'Test disabled')
+disabled_test = unittest.skipIf(True, 'Test disabled')
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/tests/test_parfors.py]: moved

**Comment by sklam** [comment]
I should have addressed everything. I want to defer the change to the exceptions in the next PR.

**Comment by stuartarchibald** [comment]
> I should have addressed everything. I want to defer the change to the exceptions in the next PR.

Thanks, postponing changes to exceptions is fine.

**Pull request merged**